### DISCUSSION TITLE
Is there any reason not to set `core.fsmonitor = watchman` globally?

### DISCUSSION THREAD

**Discussion by emilazy** [jj-vcs/jj#3894] [Q&A] [answered]
I work with at least one repository big enough to want to use Watchman. Currently I have it set as a per‐repository configuration, but if I already have it installed and working, is there any reason I shouldn’t just set it globally so I don’t have to think about it? Can it be meaningfully slower on small repositories? Are there any potential correctness issues?

**Comment by yuja** [comment, marked as answer]
There's a know bug that `.gitignore` change can be unnoticed. Other than that, I don't see any problem by enabling watchman globally.
 
https://github.com/martinvonz/jj/issues/2613

**Comment by arxanas** [comment]
Watchman may consume operating system resources (like inotify watches). In those cases, you may exhaust all of your OS file-watching resources if you watch too many paths. Watchman offers the `relative_root` mechanism to help alleviate this (particularly useful if all of your repositories happen to live in one directory): https://facebook.github.io/watchman/docs/file-query#relative-roots

I've had non-Watchman tooling start a development server and subscribe to inotify directly, which I guess wasn't cleaned up properly and then exhausted the global resource and caused all file-watching to stop working, although I think Watchman generally behaves better with respect to cleaning up resources.

**Reply by emilazy** [reply]
Thank you for the detailed reply! It looks like the macOS FSEvents API has a limit of 4096 watched paths, but Nixpkgs has *way* more files than that and seems to work fine, so either it’s silently failing to actually use it or a whole subtree counts as a single path or something. Does Jujutsu support using `relative_root`?

**Reply by arxanas** [reply]
"Watched paths" might really mean "watched roots"? I'm not personally familiar with FSEevents, so I can't say.

I don't recall for certain if the Watchman integration uses relative roots, but I believe it does.
